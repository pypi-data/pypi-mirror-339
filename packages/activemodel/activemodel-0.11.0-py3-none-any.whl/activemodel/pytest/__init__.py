from .transaction import database_reset_transaction
from .truncate import database_reset_truncate
