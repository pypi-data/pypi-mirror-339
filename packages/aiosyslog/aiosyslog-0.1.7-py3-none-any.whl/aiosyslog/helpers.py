from datetime import datetime


def datetime2rfc3339(dt, is_utc=False):
    if is_utc is False:
        # calculating timezone
        d1 = datetime.now()
        d2 = datetime.utcnow()
        diff_hr = (d1 - d2).seconds / 60 / 60
        tz = ""

        if diff_hr == 0:
            tz = "Z"
        else:
            if diff_hr > 0:
                tz = "+%s" % (tz)

            tz = "%s%.2d%.2d" % (tz, diff_hr, 0)

        return "%s%s" % (dt.strftime("%Y-%m-%dT%H:%M:%S.%f"), tz)

    else:
        return dt.isoformat() + 'Z'
