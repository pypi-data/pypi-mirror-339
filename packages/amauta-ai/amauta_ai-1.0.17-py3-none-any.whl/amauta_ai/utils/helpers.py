"""Helper utility functions for AMAUTA.

This module provides various utility functions used throughout the AMAUTA project.
"""

import os
import sys
import time
from concurrent.futures import ThreadPoolExecutor
from typing import Any, Callable, Dict, List, Optional, TypeVar, Union

import typer
from rich.console import Console
from rich.progress import Progress, SpinnerColumn, TextColumn
from rich.table import Table

from amauta_ai.utils.logger import get_logger

# Setup logger
logger = get_logger(__name__)

# Type variables for generic functions
T = TypeVar("T")


def is_offline_mode() -> bool:
    """Check if application is running in offline mode.
    
    Offline mode is activated by setting the AMAUTA_OFFLINE environment variable
    or using the --offline flag in CLI commands.
    
    Returns:
        bool: True if offline mode is active, False otherwise
    """
    return os.environ.get("AMAUTA_OFFLINE", "0").lower() in ("1", "true", "yes")


def is_debug_mode() -> bool:
    """Check if application is running in debug mode.
    
    Debug mode is activated by setting the AMAUTA_DEBUG environment variable.
    
    Returns:
        bool: True if debug mode is active, False otherwise
    """
    return os.environ.get("AMAUTA_DEBUG", "0").lower() in ("1", "true", "yes")


def create_spinner(text: str) -> Progress:
    """Create a spinner for displaying progress during long-running operations.
    
    Args:
        text (str): The text to display next to the spinner
        
    Returns:
        Progress: A rich Progress object with a spinner configuration
    """
    return Progress(
        SpinnerColumn(),
        TextColumn(f"[bold blue]{text}[/bold blue]"),
        transient=True,
    )


def run_with_spinner(func: Callable[..., T], text: str, *args: Any, **kwargs: Any) -> T:
    """Run a function with a spinner indicating progress.
    
    Args:
        func: The function to run
        text: The text to display next to the spinner
        *args: Arguments to pass to the function
        **kwargs: Keyword arguments to pass to the function
        
    Returns:
        The return value of the function
    """
    spinner = create_spinner(text)
    with spinner:
        spinner.start()
        try:
            result = func(*args, **kwargs)
            return result
        finally:
            spinner.stop()


def create_table(title: str, columns: List[str]) -> Table:
    """Create a rich formatted table.
    
    Args:
        title (str): The title of the table
        columns (List[str]): The column names
        
    Returns:
        Table: A rich Table object
    """
    table = Table(title=title)
    for column in columns:
        table.add_column(column, style="cyan", no_wrap=True)
    return table


def print_info(message: str) -> None:
    """Print an informational message.
    
    Args:
        message (str): The message to print
    """
    console = Console()
    console.print(f"[bold blue]Info:[/bold blue] {message}")


def print_success(message: str) -> None:
    """Print a success message.
    
    Args:
        message (str): The message to print
    """
    console = Console()
    console.print(f"[bold green]Success:[/bold green] {message}")


def print_warning(message: str) -> None:
    """Print a warning message.
    
    Args:
        message (str): The message to print
    """
    console = Console()
    console.print(f"[bold yellow]Warning:[/bold yellow] {message}")


def print_error(message: str) -> None:
    """Print an error message.
    
    Args:
        message (str): The message to print
    """
    console = Console()
    console.print(f"[bold red]Error:[/bold red] {message}")


def execute_in_parallel(func: Callable[[Any], T], items: List[Any], max_workers: int = None) -> List[T]:
    """Execute a function in parallel for multiple items.
    
    Args:
        func (callable): The function to execute
        items (list): The items to process
        max_workers (int, optional): Maximum number of worker threads
        
    Returns:
        List[T]: The list of results
    """
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        return list(executor.map(func, items))


def confirm_action(message: str, default: bool = False) -> bool:
    """Ask for user confirmation before proceeding with an action.
    
    Args:
        message (str): The confirmation message
        default (bool): Default response if user just presses Enter
        
    Returns:
        bool: True if confirmed, False otherwise
    """
    return typer.confirm(message, default=default)


def select_option(message: str, options: List[str], default: Optional[str] = None) -> str:
    """Ask user to select an option from a list.
    
    Args:
        message (str): The prompt message
        options (List[str]): Available options
        default (str, optional): Default option
        
    Returns:
        str: Selected option
    """
    return typer.prompt(
        message,
        type=typer.Choice(options),
        default=default or options[0],
        show_choices=True,
    )


def get_input(prompt: str, default: Optional[str] = None) -> str:
    """Get user input with an optional default value.
    
    Args:
        prompt (str): The prompt message
        default (str, optional): Default value if user just presses Enter
        
    Returns:
        str: User input
    """
    return typer.prompt(prompt, default=default or "") 