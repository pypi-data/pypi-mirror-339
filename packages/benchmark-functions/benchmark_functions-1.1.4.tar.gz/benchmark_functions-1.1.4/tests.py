#!/usr/bin/env python3

import unittest as ut
import sys, logging
from benchmark_functions import benchmark_functions as bf

DIFFERENCE_THRESHOLD = 1e-6

FUNCTIONS_TO_TEST = [ # parametric functions are excluded, they will be tested separately
	bf.Hypersphere, 
	bf.Hyperellipsoid, 
	bf.Ackley,
	bf.Rosenbrock,
	bf.Rastrigin,
	bf.Schwefel,
	bf.Griewank,
	bf.Ackley,
	bf.Michalewicz,
	bf.EggHolder, 				# <- bad for t<=1e-7, good for t>=1e-6
	bf.Keane,
	bf.Rana, 							# <- bad for t<=1e-7, good for t>=1e-6
	bf.Easom,
	bf.DeJong5,
	bf.GoldsteinAndPrice,
	bf.PichenyGoldsteinAndPrice,
	bf.StyblinskiTang,
	bf.McCormick,
	bf.MartinGaddy,
	bf.Schaffer2,
	bf.Himmelblau,
	# bf.PitsAndHoles # <- it takes too long
	]
# DeJong3 tested separately (too expensive producing all the optima)


def within_bounds(x, lb, ub):
	return x>=lb and x<=ub

class TestMetadata(ut.TestCase):
	def test_loading(self):
		for c in FUNCTIONS_TO_TEST:
			f = c()
			try:
				f.load_info()
			except Exception as e:
				self.fail(f"In loading metadata of function {f.name()}, found error {e}")		
	# a simple test to see if all the references load properly
	def test_all_references(self):
		for c in FUNCTIONS_TO_TEST:
			f = c()
			try:
				f.reference()
			except Exception as e:
				self.fail(f"In loading the reference of function {f.name()} found error {e}")

class TestOptima(ut.TestCase):
	def test_minima(self):
		log = logging.getLogger("TestOptima")
		for c in FUNCTIONS_TO_TEST:
			f=c()
			f.load_info()
			n_minima = f.n_minima()
			if '*' in n_minima:
				del n_minima['*']
				for i in range(1,11):
					if not i in n_minima:
						n_minima[i]=1
			for k in n_minima:
				if k==2:
					f=c() # this is a workaround for functions defined only for 2 dimensions
				else:
					f=c(n_dimensions=k)
				log.debug(f"Testing minima of {f.name()} for D={k}")
				for m in f.minima():
					res = f.testLocalMinimum(m.position,strict=m.region_type == 'convex', score_threshold=DIFFERENCE_THRESHOLD)
					if not res[0]:
						self.fail(f"In function {f.name()} local minimum {m} is invalid: {res[2]}")
						
	def test_DeJong3(self):
		log = logging.getLogger("TestOptima")
		for k in range(1,4):
			f = bf.DeJong3(n_dimensions=k)
			log.debug(f"Testing minima of {f.name()} for D={k}")
			minima = f.minima()
			self.assertEqual(len(minima),pow(8,k))
			for m in minima:
				lb, ub = f.suggested_bounds()
				for i in range(k):
					self.assertTrue(within_bounds(m[i], lb[i], ub[i]))
				res = f.testLocalMinimum(m.position,strict=m.region_type == 'convex', score_threshold=DIFFERENCE_THRESHOLD)
				if not res[0]:
					self.fail(f"In function {f.name()} local minimum {m} is invalid: {res[2]}")



if __name__=='__main__':
	logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)
	ut.main()
