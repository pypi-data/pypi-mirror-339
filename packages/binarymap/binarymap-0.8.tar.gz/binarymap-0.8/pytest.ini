[pytest]
addopts =
  binarymap
  docs
  --nbval
  --ignore setup.py
  --ignore docs/_build
  --doctest-modules
