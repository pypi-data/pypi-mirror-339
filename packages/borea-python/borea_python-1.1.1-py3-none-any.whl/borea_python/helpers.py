import re
from typing import Dict, Union

import click


class Helpers:
    @classmethod
    def sanitize_string(cls, s: str) -> str:
        """Replace dash, slash, period, comma, pipe, colon, semicolon, and space delimiters with underscore"""
        s = re.sub(r"[-/.,|:; ]", "_", s)
        # Remove all other special characters (keeping alphanumerics and underscores)
        s = re.sub(r"[^\w]", "", s)
        return s

    @classmethod
    def sanitize_and_lower_string(cls, s: str) -> str:
        return cls.sanitize_string(s).lower()

    @classmethod
    def clean_lower(cls, tag: str) -> str:
        """Clean tag name to be a valid Python identifier"""
        return cls.sanitize_and_lower_string(tag)

    @classmethod
    def clean_capitalize(cls, name: str) -> str:
        """Clean name to be a valid Python identifier"""
        # Remove spaces and dashes, convert to camel case
        name = cls.sanitize_string(name)
        words = name.split("_")
        return "".join(word if word.isupper() else word.capitalize() for word in words)

    @classmethod
    def clean_parameter_name(cls, name: str) -> str:
        """Clean parameter name to be a valid Python identifier"""
        return cls.sanitize_and_lower_string(name)

    @classmethod
    def clean_type_name(cls, type_name: str) -> str:
        """Clean type name to be a valid Python type"""
        if "int" in type_name:
            return "int"
        type_map = {
            "string": "str",
            "integer": "int",
            "boolean": "bool",
            "number": "float",
            "array": "List",
            "object": "Dict[str, Any]",
        }
        return type_map.get(type_name.lower(), type_name)

    @classmethod
    def clean_file_name(cls, name: str) -> str:
        """Clean name to be a valid file name"""
        return cls.sanitize_and_lower_string(name)

    @classmethod
    def clean_schema_name(cls, name: str) -> str:
        """Clean name to be a valid Python identifier"""
        return cls.sanitize_and_lower_string(name)

    @classmethod
    def format_description(cls, description: str) -> str:
        """Format description to be a valid Python docstring"""
        return (
            description.replace("\\t", "")
            .replace("\\n", "")
            .replace("\\r", "")
            .replace('"', "'")
        )

    @classmethod
    def format_type(cls, type_info: Union[Dict, str, None]) -> str:
        resolved_type: str = "Any"
        schema_type: str = resolved_type
        type_is_schema: bool = False

        def set_schema_type_to_resolved_type(value: str):
            nonlocal resolved_type
            nonlocal schema_type
            resolved_type = value
            schema_type = value

        if type_info is None:
            pass
        elif isinstance(type_info, str) and type_info not in ["object", "array"]:
            set_schema_type_to_resolved_type(cls.clean_type_name(type_info))
        elif isinstance(type_info, dict):
            type_is_schema = type_info.get("type_is_schema", False)
            if "$ref" in type_info:
                set_schema_type_to_resolved_type(type_info["$ref"].split("/")[-1])
            elif type_info.get("type") == "array":
                items = type_info.get("items", {})
                items_type = cls.format_type(items)[0]
                resolved_type = f"List[{items_type}]"
                schema_type = items_type
                type_is_schema = items.get("type_is_schema", False)
            elif "type" in type_info and type_info["type"] not in ["object", "array"]:
                set_schema_type_to_resolved_type(cls.clean_type_name(type_info["type"]))
            elif "allOf" in type_info:
                # TODO fix this when it hits
                # not hitting...
                value = " & ".join(
                    (
                        cls.format_type(item)[0]
                        if isinstance(item, dict) and "$ref" not in item
                        else item["$ref"].split("/")[-1]
                    )
                    for item in type_info["allOf"]
                )
                set_schema_type_to_resolved_type(value)
            elif "oneOf" in type_info:
                # not hitting...
                value = f"Union[{', '.join(cls.format_type(item)[0] for item in type_info['oneOf'])}]"
                set_schema_type_to_resolved_type(value)
            elif "anyOf" in type_info:
                # not hitting...
                value = f"Union[{', '.join(cls.format_type(item)[0] for item in type_info['anyOf'])}]"
                set_schema_type_to_resolved_type(value)
            elif "not" in type_info:
                # not hitting...
                set_schema_type_to_resolved_type("Any")
            else:
                pass
        else:
            pass

        return [resolved_type, schema_type, type_is_schema]

    @staticmethod
    def run_ruff_on_path(path: str):
        import subprocess

        def ruff(*args: str):
            subprocess.run(
                ["ruff", *args],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )

        try:
            # TODO: get generated code good enough to be fixed
            # Run ruff check (linting)
            ruff("check", "--fix", path)

            # Run ruff format (formatting)
            ruff("format", path)

        except subprocess.CalledProcessError as e:
            click.echo(f"Error running ruff: {e}")
