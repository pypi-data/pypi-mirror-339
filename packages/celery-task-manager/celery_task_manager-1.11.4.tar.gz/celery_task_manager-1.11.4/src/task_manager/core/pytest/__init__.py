from .fixtures import *  # NOQA F401
