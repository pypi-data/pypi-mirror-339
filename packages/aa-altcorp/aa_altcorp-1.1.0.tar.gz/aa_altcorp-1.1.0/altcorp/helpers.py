"""
Helper functions
"""

# Third Party
from dhooks_lite import Embed, Webhook
from discord import Embed as DiscordEmbed

# Django
from django.contrib.auth.models import User
from django.core.exceptions import ObjectDoesNotExist

# Alliance Auth
from allianceauth.eveonline.models import EveCharacter
from allianceauth.framework.api.evecharacter import get_user_from_evecharacter
from allianceauth.services.hooks import get_extension_logger
from esi.models import Token

from .app_settings import AC_REQUIRED_SCOPES, AC_WEBHOOK, discord_bot_active

if discord_bot_active():
    # Third Party
    from aadiscordbot import tasks as bot_tasks  # pylint: disable=import-error


logger = get_extension_logger(__name__)


def member_tokens_count_for_corp(
    corp_id: int,
    quick_check: bool = False,
) -> tuple[int, list[str]]:
    """
    Returns the number of valid tokens for a given corp and a list of names without valid tokens.
    """
    members = EveCharacter.objects.filter(corporation_id=corp_id)
    invalid_names = []  # List to store names of members without valid tokens
    valid_count = 0  # Counter for valid tokens

    # Iterate over members and process each
    for member in members:
        if has_required_scopes_for_request(member, quick_check=quick_check):
            valid_count += 1  # Increment count if the token is valid
        else:
            invalid_names.append(
                member.character_name
            )  # Add name to the list if token is invalid
    return valid_count, invalid_names


def has_required_scopes_for_request(
    char: EveCharacter,
    quick_check: bool = False,
) -> bool:
    """Returns True if the given character has all required scopes, even across multiple tokens.

    Params:
    - char: EveCharacter object.
    - quick_check: if True, skips token validity checks to save time.
    """
    try:
        user = get_user_from_evecharacter(character=char)
        state_name = user.profile.state.name
    except ObjectDoesNotExist:
        return False

    if state_name not in AC_REQUIRED_SCOPES:
        return False

    required_scopes = set(AC_REQUIRED_SCOPES[state_name])

    token_qs = Token.objects.filter(character_id=char.character_id)
    if not quick_check:
        token_qs = token_qs.require_valid()

    # Accumulate all granted scope names across all tokens
    granted_scopes = set()
    for token in token_qs:
        granted_scopes.update(token.scopes.values_list("name", flat=True))

    # Check if all required scopes are present
    return required_scopes.issubset(granted_scopes)


def send_embed_message(
    webhook: bool = False, user: User = None, embed: Embed = None, cont: str = ""
):
    """Sends an embed message via webhook or user DM."""

    if webhook:
        hook = Webhook(
            AC_WEBHOOK,
            username="Alt Corp Alert",
        )
        hook.execute(embeds=[embed], content=cont)

    if user is not None:
        if discord_bot_active():
            e2 = DiscordEmbed.from_dict(embed.asdict())
            bot_tasks.send_message(user=user, embed=e2)
