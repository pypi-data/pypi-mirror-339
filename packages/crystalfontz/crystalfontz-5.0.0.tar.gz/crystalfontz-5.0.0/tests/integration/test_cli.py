#!/usr/bin/env bash

import pytest

from tests.helpers import Cli


def test_backlight_contrast(cli: Cli, confirm) -> None:
    cli("backlight", "0.2")
    cli("contrast", "0.4")

    confirm("Did the backlight and contrast settings change?")


def test_send(cli: Cli, confirm) -> None:
    cli("send", "0", "0", "Hello world!")

    confirm('Did the LCD display "Hello world!"?')


def test_line(cli: Cli, confirm) -> None:
    cli("line", "1", "Line 1")
    cli("line", "2", "Line 2")

    confirm('Does the LCD display "Line 1" and "Line 2"?')


def test_clear(cli: Cli, confirm) -> None:
    cli("clear")

    confirm("Did the LCD clear?")


def test_cursor(cli: Cli, confirm) -> None:
    cli("cursor", "position", "1", "3")
    cli("cursor", "style", "BLINKING_BLOCK")

    confirm("Did the cursor move and start blinking?")


def test_ping(cli: Cli) -> None:
    pong: bytes = cli("ping", "pong").stdout.strip()

    assert pong == b"pong"


def test_status(cli: Cli, snapshot) -> None:
    assert cli("status").stdout.strip() == snapshot


def test_versions(cli: Cli, snapshot) -> None:
    assert cli("versions").stdout.strip() == snapshot


def test_reboot(cli: Cli, confirm) -> None:
    cli("--timeout", "1.0", "power", "reboot-lcd")

    confirm("Did the LCD reboot?")


@pytest.mark.skip
def test_detect() -> None:
    raise NotImplementedError("test_detect")


def test_listen(cli, confirm) -> None:
    with cli.bg("listen"):
        confirm("Mash some buttons. Are events showing up?")


def test_listen_for(cli: Cli) -> None:
    cli("listen", "--for", "1.0")


def test_marquee(cli, confirm) -> None:
    with cli.bg("effects", "marquee", "0", "Josh is cool"):
        confirm("Is the LCD showing a marquee effect?")


def test_marquee_for(cli: Cli) -> None:
    cli("effects", "--for", "1.0", "marquee", "1", "Josh is cool")


def test_screensaver(cli, confirm) -> None:
    with cli.bg("effects", "screensaver", "Josh!"):
        confirm("Is the LCD showing a screensaver effect?")


def test_screensaver_for(cli: Cli) -> None:
    cli("effects", "--for", "1.0", "screensaver", "Josh!")


def test_dance_party(cli, confirm) -> None:
    cli("clear")
    cli("send", "0", "0", "Carameldansen!!")
    with cli.bg("effects", "dance-party"):
        confirm("Is the LCD showing a dance party effect?")


@pytest.mark.skip
def test_read_user_flash() -> None:
    raise NotImplementedError("test_read_user_flash")


@pytest.mark.skip
def test_poll_keypad() -> None:
    raise NotImplementedError("test_poll_keypad")
