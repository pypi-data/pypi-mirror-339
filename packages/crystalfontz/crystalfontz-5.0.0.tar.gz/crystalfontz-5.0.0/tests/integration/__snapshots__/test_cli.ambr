# serializer version: 1
# name: test_status
  b'CFA533 Status:\n--------------\nTemperature sensors enabled: \nKey states:\n  up: keypress=KP_UP, pressed=no, pressed_since=yes, released_since=yes\n  enter: keypress=KP_ENTER, pressed=no, pressed_since=yes, released_since=yes\n  exit: keypress=KP_EXIT, pressed=no, pressed_since=yes, released_since=yes\n  left: keypress=KP_LEFT, pressed=no, pressed_since=yes, released_since=yes\n  right: keypress=KP_RIGHT, pressed=no, pressed_since=yes, released_since=yes\n  down: keypress=KP_DOWN, pressed=no, pressed_since=yes, released_since=yes\nATX Power Switch Functionality Settings:\n  Functions enabled: \n  Auto-Polarity Enabled: no\n  Reset Inverted: no\n  Power Inverted: no\n  Power Pulse Length (seconds): None\nWatchdog Counter: 0\nContrast: 0.3137254901960784\nContrast (CFA633 Compatible): 0.24\nBacklight:\n  Keypad Brightness: 0.2\n  LCD Brightness: 0.2'
# ---
# name: test_versions
  b'CFA533: h1.4, u1v2'
# ---
